### PR TITLE
restart via API allows to copy description

### LINKED ISSUE #626: Restarting a entry- missing description
Hi Kevin,

we use the restart function to create a new entry based on an existing entry but the new entry doesn't contain the description of the restarted entry.

Is this a bug or should it be a feature?

### PR DESCRIPTION
**Pull request by kevinpapst**
## Description

- Moved the logic to restart a timesheet entry to the API
- Added option to copy description and tags

New behavior:
- A "recent activity" restart will restart the record WITHOUT copying the data
- A restart from the timesheet will copy description & tags

Fixes #626

Can be tested at: https://demo-branch.kimai.org/en/timesheet/

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I ran `bin/console kimai:codestyle --fix` to verify the correct code style
- [x] I have added tests to cover my changes
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

### CODE REVIEW

**Comment by blochwitz** [comment]
Cant review that while during the installation process creating the database schema fails:

```
Schema-Tool failed with Error 'An exception occurred while executing 'CREATE TABLE kimai2_tags (id INT AUTO_INCREMENT NOT NULL, name VARCHAR(255) NOT NULL, UNIQUE INDEX UNIQ_27CAF  
  54C5E237E06 (name), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci ENGINE = InnoDB': 
```

**Comment by kevinpapst** [comment]
Hm, the installation works on a lot of machines with MySQL and MariaDB. 
Just copy&pasted that into two sql consoles just to make sure, worked as well.
Please open a new issue with the full error message.

**Comment by kevinpapst** [comment]
For simple testing, I installed the branch at: https://demo-branch.kimai.org/en/timesheet/

**Comment by blochwitz** [comment]
Hmm, i made a clean installation of the branch following the installation instruction and got the error during the installation step creating the database schema.

In the demo the new action „Start again“ works fine...

**Comment by kevinpapst** [comment]
Thanks for testing! 
I am sure that you got the error, I just don't know what the reason might be as there was no error message visible. If you have the time we should investigate a bit, I don't want to risk some hidden bugs.

File: src/API/TimesheetController.php [hunk 1]
```diff
@@ -527,4 +530,91 @@ public function stopAction($id)
...
+        }
+
+        if ($timesheet->getUser()->getId() !== $user->getId() && !$this->isGranted('start_other_timesheet')) {
+            throw new AccessDeniedHttpException('You are not allowed to re-start this timesheet');
+        }
+
+        if (!$this->isGranted('start', $timesheet)) {
+            throw new AccessDeniedHttpException('You are not allowed to re-start this timesheet');
+        }
+
```
Comment (kevinpapst) [inline src/API/TimesheetController.php]: empty line

File: src/API/TimesheetController.php [hunk 2]
```diff
@@ -527,4 +530,91 @@ public function stopAction($id)
...
+        /** @var Timesheet $timesheet */
+        $timesheet = $this->repository->find($id);
+        /** @var User $user */
+        $user = $this->getUser();
+
+        if (null === $timesheet) {
+            throw new NotFoundException();
+        }
+
+        if ($timesheet->getUser()->getId() !== $user->getId() && !$this->isGranted('start_other_timesheet')) {
```
Comment (kevinpapst) [inline src/API/TimesheetController.php]: unneccessary, is covered in the voter next call `if (!$this->isGranted('start', $timesheet)) {`

**Comment by blochwitz** [comment]
Okay. I try to explain.

I made two tests: one with an existing database (kimai2 v0.9) by using the migration command - this fails with an error. another one with an database i've cleared before - it fails with an error too (the key part you see above).

**Comment by kevinpapst** [comment]
Can you share the full error messages for both ways? Or the database, so I can test it manually?

**Comment by blochwitz** [comment]
Okay. This is the Output after trying to migrate

Bildschirmfoto 2019-05-13 um 16 12 19

**Comment by kevinpapst** [comment]
Thats very helpful, thanks! 
I know what to do and must check if I can reconfigure my databases to show these key-length errors as well.
I'll prepare a new branch with these changes incorporated and let you know.

**Comment by rarvin** [comment]
Dear all,

Tried testing this and could confirm that everything works well.

Restarting an activity from the timesheet copy all of the details including
existing applied tages, this is great functionality.
Started a task from the + and I can enter the description from scratch,
even could edit the description for a running or stopped task.
Started a task from the recent activities and the description is not copied
which is as intended.
Could stop tasks from the play button; create new tasks from scratch.

In general the copy description functionality works very well. Thank you.

**Comment by rarvin** [comment]
Dear all,

Carried out testing using the following link

Tried testing this and could confirm that everything works well.
https://demo-branch.kimai.org/en/team/timesheet/

Restarting an activity from the timesheet copy all of the details including existing applied tages, this is great functionality.
Started a task from the + and I can enter the description from scratch, even could edit the description for a running or stopped task. 
Started a task from the recent activities and the description is not copied which is as intended.
Could stop tasks from the play button; create new tasks from scratch.

In general the copy description functionality works very well. Thank you.

**Pull request merged**